snuba/query/parser/validation/functions.py

```diff
@@ -103,34 +102,38 @@ def validate(self, exp: Expression, data_source: DataSource) -> None:
         ):
             return
 
-        entity_validators: Mapping[str, FunctionCallValidator] = {}
-        entity = None
-
-        query_entity = QueryEntityFinder().visit(data_source)
-        if query_entity:
-            entity = get_entity(query_entity.key)
-            entity_validators = entity.get_function_call_validators()
-
-        common_function_validators = (
-            entity_validators.keys() & default_validators.keys()
-        )
-        if common_function_validators:
-            logger.warning(
-                "Dataset validators are overlapping with default ones. Entity: %s. Overlap %r",
-                entity,
-                common_function_validators,
-                exc_info=True,
-            )
-
-        validators = ChainMap(default_validators, entity_validators)
+        # First do global validation, independent of entities
         try:
-            # TODO: Decide whether these validators should exist at the Dataset or Entity level
-            validator = validators.get(exp.function_name)
-            if validator is not None:
-                validator.validate(exp.parameters, data_source)
+            for validator in global_validators:
+                validator.validate(exp.function_name, exp.parameters, data_source)
         except InvalidFunctionCall as exception:
             raise InvalidExpressionException(
                 exp,
-                f"Illegal call to function {exp.function_name}: {str(exception)}",
+                f"Invalid function name {exp.function_name}: {str(exception)}",
                 report=False,
             ) from exception
+
+        # Then do entity specific validation, if there are any
+        entity = None
+        query_entities = QueryEntityFinder().visit(data_source)
+
+        for query_entity in query_entities:
+            entity = get_entity(query_entity.key)
+
+            entity_validator = entity.get_function_call_validators().get(
+                exp.function_name
+            ) or default_validators.get(exp.function_name)
```

Comment:
If you have a join query (thus with multiple entities) wouldn't you be applying the default validation multiple times if neither of the entity provides a validator for this function ?